src/adapters/supabase/helpers/log.ts

```diff
@@ -0,0 +1,196 @@
+import { getAdapters, getBotContext } from "../../../bindings";
+import { Payload } from "../../../types";
+import { getOrgAndRepoFromPath } from "../../../utils/private";
+interface Log {
+  repo: string | null;
+  org: string | null;
+  commentId: number | undefined;
+  issueNumber: number | undefined;
+  logMessage: string;
+  errorType: levels;
+  timestamp: string;
+}
+
+export enum levels {
+  error = 0,
+  warn = 1,
+  info = 2,
+  http = 3,
+  verbose = 4,
+  debug = 5,
+  silly = 6,
+}
+
+export class GitHubLogger {
+  private supabase = getAdapters().supabase;
+  private maxLevel;
+  private app;
+  private logEnvironment;
+  private logQueue: Log[] = []; // Your log queue
+  private maxConcurrency = 5; // Maximum concurrent requests
+  private retryLimit = 3; // Maximum number of retries
+  private retryDelay = 1000; // Delay between retries in milliseconds
+  private throttleCount = 0;
+
+  constructor(app: string, logEnvironment: string, maxLevel: keyof typeof levels) {
+    this.app = app;
+    this.logEnvironment = logEnvironment;
+    this.maxLevel = levels[maxLevel];
+  }
+
+  async sendLogsToSupabase({ repo, org, commentId, issueNumber, logMessage, errorType, timestamp }: Log) {
+    try {
+      const { error } = await this.supabase.from("log_entries").insert([
+        {
+          repo_name: repo,
+          level: errorType,
+          org_name: org,
+          comment_id: commentId,
+          log_message: logMessage,
+          issue_number: issueNumber,
+          timestamp,
+        },
+      ]);
+
+      if (error) {
+        console.error("Error logging to Supabase:", error.message);
+        return;
+      }
+    } catch (error) {
+      if (error instanceof Error) {
+        console.error("An error occurred:", error.message);
+      }
+    }
+  }
+
+  async processLogs(log: Log) {
+    try {
+      await this.sendLogsToSupabase(log);
+    } catch (error) {
+      console.error("Error sending log, retrying:", error);
+      await this.retryLog(log);
+    }
+  }
+
+  async retryLog(log: Log, retryCount = 0) {
+    if (retryCount >= this.retryLimit) {
+      console.error("Max retry limit reached for log:", log);
+      return;
+    }
+
+    await new Promise((resolve) => setTimeout(resolve, this.retryDelay));
+
+    try {
+      await this.sendLogsToSupabase(log);
+    } catch (error) {
+      console.error("Error sending log (after retry):", error);
+      await this.retryLog(log, retryCount + 1);
+    }
+  }
+
+  async processLogQueue() {
+    while (this.logQueue.length > 0) {
```

Comment:
@devpanther Why do we need this loq queue? Why can't we simply send logs to supabase every time there is a new log and let the node's event loop handle everything?

Reply:
> @devpanther Why do we need this loq queue? Why can't we simply send logs to supabase every time there is a new log and let the node's event loop handle everything?

I just don't want to overwhelm the db at any point.. right now it logs go in one at a time, but its to handle cases where they come in bulk.

I've seen supabase db lose connection when too many requests is sent, so I am handling it in queue and retrying on error to make sure every single log get in and in their right order

Reply:
I think this is a more robust approach, good job!